url20b0f880a5_jgraph_legacy_jgraph5_tgz-pJ8-org_jgraph_example_HelloWorldJ8/src/com/mxgraph/view/mxGraph.java

```diff
@@ -0,0 +1,7565 @@
+
+		return dirty;
+	}
+
+	/**
+	 * Extends the canvas by doing another validation with a shifted
+	 * global translation if the bounds of the graph are below (0,0).
+	 * 
+	 * The first validation is required to compute the bounds of the graph
+	 * while the second validation is required to apply the new translate.
+	 */
+	protected void updateOrigin()
+	{
+		mxRectangle bounds = getGraphBounds();
+
+		if (bounds != null)
+		{
+			double scale = getView().getScale();
+			double x = bounds.getX() / scale - getBorder();
+			double y = bounds.getY() / scale - getBorder();
+
+			if (x < 0 || y < 0)
+			{
+				double x0 = Math.min(0, x);
+				double y0 = Math.min(0, y);
+
+				origin.setX(origin.getX() + x0);
+				origin.setY(origin.getY() + y0);
+
+				mxPoint t = getView().getTranslate();
+				getView().setTranslate(
+						new mxPoint(t.getX() - x0, t.getY() - y0));
+			}
+			else if ((x > 0 || y > 0)
+					&& (origin.getX() < 0 || origin.getY() < 0))
+			{
+				double dx = Math.min(-origin.getX(), x);
+				double dy = Math.min(-origin.getY(), y);
+
+				origin.setX(origin.getX() + dx);
+				origin.setY(origin.getY() + dy);
+
+				mxPoint t = getView().getTranslate();
+				getView().setTranslate(
+						new mxPoint(t.getX() - dx, t.getY() - dy));
+			}
+		}
+	}
+
+	/**
+	 * Returns the cells that have been removed from the model.
+	 */
+	public Object[] getRemovedCellsForChanges(List<mxUndoableChange> changes)
+	{
+		List<Object> result = new ArrayList<Object>();
+		Iterator<mxUndoableChange> it = changes.iterator();
+
+		while (it.hasNext())
+		{
+			Object change = it.next();
+
+			if (change instanceof mxRootChange)
+			{
+				break;
+			}
+			else if (change instanceof mxChildChange)
+			{
+				mxChildChange cc = (mxChildChange) change;
+
+				if (cc.getParent() == null)
+				{
+					result.addAll(mxGraphModel.getDescendants(model,
+							cc.getChild()));
+				}
+			}
+			else if (change instanceof mxVisibleChange)
+			{
+				Object cell = ((mxVisibleChange) change).getCell();
+				result.addAll(mxGraphModel.getDescendants(model, cell));
+			}
+		}
+
+		return result.toArray();
+	}
+
+	/**
+	 * Processes the changes and returns the minimal rectangle to be
+	 * repainted in the buffer. A return value of null means no repaint
+	 * is required.
+	 */
+	public mxRectangle processChanges(List<mxUndoableChange> changes,
+			boolean invalidate, boolean ignoreDirty)
+	{
+		mxRectangle bounds = null;
+		Iterator<mxUndoableChange> it = changes.iterator();
+
+		while (it.hasNext())
+		{
+			mxRectangle rect = processChange(it.next(), invalidate, ignoreDirty);
+
+			if (bounds == null)
+			{
+				bounds = rect;
+			}
+			else
+			{
+				bounds.add(rect);
+			}
+		}
+
+		return bounds;
+	}
+
+	/**
+	 * Processes the given change and invalidates the respective cached data
+	 * in <view>. This fires a <root> event if the root has changed in the
+	 * model.
+	 */
+	public mxRectangle processChange(mxUndoableChange change,
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

The cyclomatic complexity of this method is 34. By comparison, 99% of the methods in the CodeGuru reference dataset have a lower cyclomatic complexity. This indicates the method has a high number of decisions and it can make the logic difficult to understand and test. We recommend that you simplify this method or break it into multiple methods.